[Kyh196201] 4주차 To-do 리스트 Redux를 사용하여 리팩터링 하기

kyh196201 opened the pull request:
트레이너님 안녕하세요!

리뷰 부탁드립니다 :-)

TDD에 대한 감을 잡기위해서 정말 작은 단위로 하나 하나씩 커밋을 올렸는데, 이렇게 하는게 맞는지 잘 모르겠습니다.

초반에 감을 잘 못잡아서 PR을 너무 늦게올렸네요 😂

kyh196201 committed:
redux 관련 모듈 설치

- redux 환경 설정에 필요한 모듈을 설치했습니다.

kyh196201 committed:
과제 기본 구조 설정

reducer를 적용하기 전에 필요한 기본 파일을 생성하고 구조를 설정했습니다.

kyh196201 committed:
App 컴포넌트 테스트 코드 수정

- 테스트 코드 내에서 Provider를 이용하지 않고, jest.mock()을 이용하는 방법으로 수정했습니다.

kyh196201 committed:
UpdateTaskTitle action에 대한 테스트 추가

kyh196201 committed:
TaskTitle을 업데이트하는 액션을 구현

kyh196201 committed:
UpdateTaskTitle 테스트 코드 matchers 변경

- 새로 변경된 state가 업데이트된 taskTitle 값을 가지고 있는지 검사하는게 맞다고 생각되어서
toHaveProperty로 변경했습니다.

kyh196201 committed:
할 일을 추가하는 액션에 대한 테스트 작성

- 입력 중인 할 일이 있을 경우에 대한 테스트만 추가했습니다.

kyh196201 committed:
할 일을 추가하는 액션을 구현

- 현재 taskTitle, newId를 기준으로 새로운 task를 추가하는 코드만 구현했습니다.

kyh196201 committed:
할 일을 추가하는 액션에 대한 테스트 케이스 추가

- 할 일 추가 후 입력중인 값이 비워지는지 테스트하는 코드를 추가했습니다.

kyh196201 committed:
할 일을 추가하는 액션 수정

- 할 일 추가 후 taskTitle을 빈 값으로 변경하는 코드를 추가했습니다.

passwd10 commented on src/reducer.js:
```diff
@@ -0,0 +1,29 @@
+};
+
+const reducer = (state = initialState, action) => {
+  if (action.type === 'updateTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
```
if 문이 반복되는 형태네요. 어떻게 하면 if 문을 사용하지 않고 reducer를 구현할 수 있을까요?

passwd10 commented on src/App.jsx:
```diff
@@ -1,15 +1,25 @@
-  const { newId, taskTitle, tasks } = state;
+  const { newId, taskTitle, tasks } = useSelector(selector);
 
   function handleChangeTitle(event) {
```
value 값을 직접 받아서 하위 컴포넌트와의 의존성을 끊어주세요.

kyh196201 replied:
의존성을 낮추기 위해서 Input 컴포넌트에서 `event.target.value`를 onClick 함수에 전달하도록 수정했습니다.
맞게 수정한건지 궁금합니다! 😃

passwd10 commented on src/App.jsx:
```diff
@@ -1,15 +1,25 @@
-    taskTitle: '',
-    tasks: [],
-  });
+  const [state, setState] = useState(initialState);
```
상태관리는 Redux 에서 처리할 수 있도록 해주세요!

kyh196201 replied:
추가 기능 구현하면서 Redux에서 처리하도록 변경하겠습니다 😊

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,54 @@
+describe('reducer', () => {
+  describe('updateTaskTitle', () => {
+    it('taskTitle이 변경된다.', () => {
+      const action = {
+        type: 'updateTaskTitle',
+        payload: {
+          taskTitle: 'New Task',
+        },
+      };
```
action 함수를 모아놓은 파일을 만들어보세요. 그렇게하면 dispatch 와 test 에서 각각 사용할 수 있을거에요

kyh196201 replied:
넵! actions.js 파일 생성하고, 파일 내부에 action 함수를 선언했습니다

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,54 @@
+        taskTitle: '',
+      }, action);
+
+      expect(newState).toHaveProperty('taskTitle', 'New Task');
```
```javascript
expect(newState.taskTitle).toBe('New Task');
```
이렇게하면 조금 더 직관적으로 표현할 수 있을 것 같아요!

kyh196201 replied:
둘 중에 어느것으로 할지 고민하다가 `toHaveProperty`를 사용했는데 `toBe`가 더 직관적이네요 감사합니다 🙌

passwd10 reviewed (commented):
네 잘하고 있어요! 👍 
이번 과제에선 TDD 를 훈련하기 위해 테스트 커밋과 구현 커밋을 구분해서 순차적으로 올리도록 하고있어요.
숙달이 되면 명시적으로 커밋을 분리하지 않아도 될거에요!

kyh196201 committed:
updateTaskTitle, addTask에 대한 action creator 생성

kyh196201 committed:
App 컴포넌트 리팩토링 진행

- handleChangeTitle, handleClickAddTask에 dispatch를 적용했습니다.

kyh196201 committed:
addTask 테스트 케이스 수정

- 입력 중인 값이 없을 경우 테스트하는 코드를 추가했습니다.
- updateTaskTitle 테스트 코드에서 toHaveProperty를 toBe로 변경했습니다.

kyh196201 committed:
할 일을 추가하는 액션 수정

- taskTitle이 없을 경우 기존 state를 리턴하도록 수정했습니다.

kyh196201 committed:
할 일을 삭제하는 액션에 대한 테스트 코드 작성

- actions.js에 deleteTask 액션을 추가했습니다.

kyh196201 committed:
할 일을 삭제하는 action 구현

- App 컴포넌트 내부에 있는 상태 관리 관련 로직을 제거했습니다.
- 할 일을 삭제하는 로직을 dispatch를 이용해서 구현했습니다.

kyh196201 committed:
App 컴포넌트 리팩터링

- handleChangeTitle가 Input 컴포넌트로부터 전달된 event를 인자로 전달받았는데,
value를 직접 전달받을 수 있도록 수정했습니다.

kyh196201 committed:
reducer 테스트 코드 리팩터링

- addTask 테스트 코드 내부에 중복되는 로직을 제거했습니다.

kyh196201 committed:
ListContainer 컴포넌트 추가

- ListContainer 컴포넌트와 테스트 코드를 추가했습니다.

kyh196201 committed:
InputContainer 컴포넌트 추가

- InputContainer 컴포넌트와 테스트 코드를 추가했습니다.
- Page 컴포넌트 내부에서 InputContainer 컴포넌트를 렌더링하도록 변경했습니다.

kyh196201 committed:
App 컴포넌트 소스 리팩터링

- App 컴포넌트 내부의 상태 관리 로직을 모두 제거했습니다.
- Page 컴포넌트가 불필요하므로 Page 컴포넌트 내부의 컴포넌트를 App으로 옮겼습니다.

kyh196201 committed:
Page 컴포넌트 및 테스트 파일 삭제

kyh196201 committed:
InputContainer 테스트 코드 수정

- 할 일을 입력할 경우에 대한 테스트 케이스를 추가했습니다.

kyh196201 commented:
트레이너님 피드백 주신 내용 반영했고, Container 컴포넌트를 적용해서 소스 리팩토링했습니다.

아직 reducer에 있는 if문은 제거하지 못했습니다.

그리고 전체적으로 소스 리팩토링도 필요한데 먼저 PR 올렸습니다 😂

리뷰 요청드립니다!

kyh196201 committed:
Reducer 테스트 코드 리팩토링

- state를 인자로 전달하지 않았을 경우에 대한 테스트 코드를 추가했습니다.
- 정의되지 않은 action을 전달했을 경우 테스트하는 코드를 추가했습니다.

kyh196201 committed:
updateTaskTitle 함수 디폴트 파라미터 제거

kyh196201 committed:
TASKS fixture 파일 생성

kyh196201 committed:
App 컴포넌트 테스트 코드 수정

- Input과 List 컴포넌트가 렌더링되었는지 확인하는 테스트 코드를 추가했습니다.

kyh196201 committed:
List, ListContainer 테스트 코드 리팩터링

- 파일 내부에서 직접 선언한 tasks를 제거했습니다.

passwd10 commented on src/App.jsx:
```diff
@@ -1,46 +1,14 @@
+  return (
+    <div>
+      <h1>To-do</h1>
 
```
공백 지워주세요!

kyh196201 replied:
넵 불필요한 공백 제거했습니다 :-)

passwd10 commented on src/App.test.jsx:
```diff
@@ -1,14 +1,45 @@
+      useSelector.mockImplementation((selector) => selector({
+        ...initialState,
+        tasks: TASKS,
+      }));
```
useSelector는 상위에서 한번만 선언해주시고 given2 라는 라이브러리를 사용해서 테스트별로 필요한 값을 넣어주시면 좋을 것 같아요!

passwd10 commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,78 @@
+
+describe('InputContainer', () => {
+  it('Input이 보인다.', () => {
+    useSelector.mockImplementation((selector) => selector({
+      taskTitle: '',
+    }));
```
이 부분도 마찬가지로, 테스트별로 useSelector를 선언할 필요 없이 상위에서 한번만 선언하신 후에 given2 를 이용해서 lazy하게 필요한 값을 넣어주시면 될 것 같아요

passwd10 commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,78 @@
+        taskTitle: '',
+      }));
+
+      const dispatch = jest.fn();
+
+      useDispatch.mockImplementation(() => dispatch);
```
dispatch는 상단에서 한번만 선언해도 될 것 같아요!

kyh196201 replied:
넵! describe 내부에 한 번 선언하고, beforeEach에서 clear하도록 수정했습니다.

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,84 @@
+} from './actions';
+
+describe('reducer', () => {
+  context('state가 없으면', () => {
```
context로 state가 없는경우를 작성해주셨으니, state가 있는 경우에 대해서도 작성해보세요!

kyh196201 replied:
저번에 한 번 피드백 받은 내용인데 신경을 못썼네요 ㅎㅎ 감사합니다!

passwd10 commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,78 @@
+    expect(getByText('추가')).toBeInTheDocument();
+  });
+
+  describe('할 일 입력', () => {
```
describe로 감싸지 않아도 될 것 같아요

passwd10 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,84 @@
+    });
+  });
+
+  describe('정의되지 않은 액션', () => {
```
정의된 액션 vs 정의되지 않은 액션

지금과 같이 상반된 상황이 나오는 경우엔 context를 이용해서 표현해주세요!

kyh196201 replied:
이 부분도 수정 완료했습니다!

kyh196201 committed:
App 컴포넌트 내부 공백 삭제

kyh196201 committed:
reducer 리팩터링

- reducer 내부에서 if 문을 사용하지 않기 위해서 reducers 객체를 구현했습니다.

kyh196201 committed:
InputContainer 테스트 코드 리팩터링

- dispatch를 테스트 내에서 한번만 선언하도록 수정했습니다.

- beforeAll을 beforeEach로 수정했습니다.

kyh196201 committed:
reducer 테스트 코드 리팩터링

- 정의되어 있지 않은 액션을 테스트하는 코드를 describe에서 context로 변경했습니다.

- state가 있을 경우에 대한 context를 추가했습니다.

kyh196201 committed:
given2 모듈 설치 및 jest 환경 설정

kyh196201 committed:
ListContainer 컴포넌트 테스트 코들 리팩터링

- given2 라이브러리를 이용해서 중복되는 useSelector.mockImplementation을 제거했습니다.

- dispatch를 describe 내에서 한번만 선언하도록 수정했습니다.

kyh196201 committed:
InputContainer 컴포넌트 리팩터링

- given2를 이용해서 중복되는 useSelector.mockImplementation을 제거했습니다.

kyh196201 committed:
App 컴포넌트 리팩터링

- given2를 이용해서 중복되는 useSelector.mockImplementation을 제거했습니다.

kyh196201 commented:
트레이너님 피드백 주신 내용 모두 수정 완료했습니다 :-)

App, InputContainer, ListContainer 컴포넌트 테스트 코드를 `given2` 를 이용해서 리팩터링했습니다.

리뷰 부탁드립니다 🙌

passwd10 commented:
과제를 잘 진행하셨어요 👏👏 
두번째 과제도 계속 진행해보시죠!
